crane-cloud/backend
# Add Billing Metrics Endpoint

# Description

This PR adds route `/projects/{id}/billing/info` that returns cost information for a Project.
I currently added metrics coming directly from `kubecost` API, this is because of the given time we heard to pull this off and have the rest of the team moving. The way forward is to create our own cost allocation module
#### It expects the following body

- `start` - A timestamp for which metrics shown should start
- `end` - A timestamp for which metrics shown should end
- `window` - A string taking in two values `lastmonth` and `lastweek` this is still a range and overrides `start` and `end` values.
- `series` - A boolean default value is `False` if true returns series data in a certain period of time.
- `show_deployment` - A boolean default value is `False`, if true returns metrics showing resource costs for each Application in the project.
**Note:** Non of these fields is mandatory

#### Expected output is
![Screenshot from 2022-04-21 05-08-17](https://user-images.githubusercontent.com/18049550/164357401-57c1684b-9db8-419a-9d4e-bcb11fddcc9f.png)
When `series` is true
![Screenshot from 2022-04-21 05-10-39](https://user-images.githubusercontent.com/18049550/164357631-85b52a15-8b2a-4af1-a384-96705455f534.png)
When `show_deployments` is true
![Screenshot from 2022-04-21 05-12-25](https://user-images.githubusercontent.com/18049550/164357746-edc2e53c-3724-40e1-92b3-242b740498a4.png)
 
## Type of change

Please delete options that are not relevant.


## Trello Ticket ID

This Pr addresses [this](https://trello.com/c/gD9T57VL/999-billing-metrics-usage-endpoint) and [this](https://trello.com/c/gD9T57VL/999-billing-metrics-usage-endpoint) ticket

## How Can This Be Tested?

**Please note**: This only works on the `netlabs` cluster at the moment so it won't show any metrics for the other cluster because it is the only one with kubecost installed.
That said, 
- Add the `netlabs` cluster to your cranecloud app and deploy an application
- Check your projects billing information on route POST `/projects/id/billing/info`
- Since our app is new, it may not have enough data to visualise.
- You can comment out line `656` in `/controllers/project.py` file to test with a more active namespace like `liqo`.

I have added some TODO comments that arose from some of the following:
1. Ingress creation causing errors; I suppose this is because the new cluster has an updated Kubernetes version and the `extensions` API group of Kubernetes is getting reducted.
2. I added the `kubecost` url directly in the code; this is because `kubecost` is currently a temporary setup and the actual thing to do would be to add the URL to the cluster DB models. which might be redundant. but you can tell me what you think
3. Have the `start` field begin from the last payment date. these are currently not available so I put a pin on it.

## Checklist:
